cmd/pitr/recoverer/recoverer.go

```diff
@@ -0,0 +1,283 @@
+		flags = ` --stop-datetime="` + r.recoverTime + `"`
+
+		format := "2006-01-02 15:04:05"
+		t, err := time.Parse(format, r.recoverTime)
+		if err != nil {
+			return errors.Wrap(err, "parse date")
+		}
+		endTime = t
+	case Latest:
+	default:
+		return errors.New("wrong recover type")
+	}
+
+	for _, binlog := range r.binlogs {
+		log.Println("working with", binlog)
+
+		if r.recoverType == Date {
+			binlogArr := strings.Split(binlog, "_")
+			if len(binlogArr) < 2 {
+				return errors.New("get timestamp from binlog name")
+			}
+			binlogTime, err := strconv.ParseInt(binlogArr[1], 10, 64)
+			if err != nil {
+				return errors.Wrap(err, "get binlog time")
+			}
+			if binlogTime > endTime.Unix() {
+				return nil
+			}
+		}
+
+		binlogObj, err := r.storage.GetObject(binlog)
+		if err != nil {
+			return errors.Wrap(err, "get obj")
+		}
+
+		err = os.Setenv("MYSQL_PWD", os.Getenv("PXC_PASS"))
+		if err != nil {
+			return errors.Wrap(err, "set mysql pwd env var")
+		}
+
+		cmdString := "mysqlbinlog" + flags + " - | mysql -h" + host + " -u" + r.pxcUser
+		cmd := exec.Command("sh", "-c", cmdString)
+
+		cmd.Stdin = binlogObj
+		var outb, errb bytes.Buffer
+		cmd.Stdout = &outb
+		cmd.Stderr = &errb
+		err = cmd.Run()
+		if err != nil {
+			return errors.Wrap(err, "cmd run")
+		}
+
+		if errb.Bytes() != nil {
+			log.Println(errors.Errorf("cmd error: %s, stdout: %s", errb.String(), outb.String()))
+		}
+	}
+
+	return nil
+}
+
+func getLastBackupGTID(infoObj io.Reader) (string, error) {
+	sep := []byte("GTID of the last")
+
+	content, err := ioutil.ReadAll(infoObj)
+	if err != nil {
+		return "", errors.Wrap(err, "read object")
+	}
+
+	startIndex := bytes.Index(content, sep)
+	if startIndex == -1 {
+		return "", errors.New("no gtid data in backup")
+	}
+	newOut := content[startIndex+len(sep):]
+	e := bytes.Index(newOut, []byte("'\n"))
+	if e == -1 {
+		return "", errors.New("can't find gtid data in backup")
+	}
+	content = content[:e]
+
+	se := bytes.Index(newOut, []byte("'"))
+	set := newOut[se+1 : e]
+
+	return string(set), nil
+}
+
+func (r *Recoverer) setBinlogs() error {
+	list := r.storage.ListObjects(r.s3Prefix + "/binlog_")
+	binlogs := []string{}
+	for _, binlog := range reverseArr(list) {
+		binlogs = append(binlogs, binlog)
+
+		infoObj, err := r.storage.GetObject(r.s3Prefix + binlog + "-gtid-set")
+		if err != nil {
+			return errors.Wrap(err, "get object with gtid set")
+		}
+		content, err := ioutil.ReadAll(infoObj)
+		if err != nil {
+			return errors.Wrap(err, "get object")
+		}
+
+		isSubset, err := r.db.IsGTIDSubset(r.startGTID, string(content))
+		if err != nil {
+			return errors.Wrapf(err, "is '%s' subset of '%s", r.startGTID, string(content))
+		}
+
+		if isSubset {
+			break
+		}
+	}
+
+	r.binlogs = reverseArr(binlogs)
+
+	return nil
+}
+
+func reverseArr(arr []string) []string {
+	newArr := make([]string, len(arr))
+	for k, v := range arr {
+		newKey := len(newArr) - k - 1
+		newArr[newKey] = v
+	}
+	return newArr
+}
```

Comment:
you can reverse slice in place, without creating a new one. See https://github.com/golang/go/wiki/SliceTricks#reversing

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/673/commits/059533512393f8112ca47ab30263ab832efbaef8